sources/Engine.ts

```diff
@@ -150,54 +157,32 @@ export class Engine {
     if (typeof definition === `undefined`)
       throw new UsageError(`This package manager (${packageManager}) isn't supported by this corepack build`);
 
-    let lastKnownGoodFile = await getLastKnownGoodFile(`r+`).catch(err => {
-      if ((err as NodeError)?.code !== `ENOENT` && (err as NodeError)?.code !== `EROFS`) {
-        throw err;
-      }
-    });
-    try {
-      const lastKnownGood = lastKnownGoodFile == null || await getJSONFileContent(lastKnownGoodFile!);
-      const lastKnownGoodForThisPackageManager = getLastKnownGoodFromFileContent(lastKnownGood, packageManager);
-      if (lastKnownGoodForThisPackageManager)
-        return lastKnownGoodForThisPackageManager;
-
-      if (process.env.COREPACK_DEFAULT_TO_LATEST === `0`)
-        return definition.default;
-
-      const reference = await corepackUtils.fetchLatestStableVersion(definition.fetchLatestFrom);
-
-      try {
-        lastKnownGoodFile ??= await createLastKnownGoodFile();
-        await activatePackageManagerFromFileHandle(lastKnownGoodFile, lastKnownGood, {
-          name: packageManager,
-          reference,
-        });
-      } catch {
-        // If for some reason, we cannot update the last known good file, we
-        // ignore the error.
-      }
+    const lastKnownGood = await getLastKnownGood();
+    const lastKnownGoodForThisPackageManager = getLastKnownGoodFromFileContent(lastKnownGood, packageManager);
+    if (lastKnownGoodForThisPackageManager)
+      return lastKnownGoodForThisPackageManager;
 
-      return reference;
-    } finally {
-      await lastKnownGoodFile?.close();
-    }
-  }
+    if (process.env.COREPACK_DEFAULT_TO_LATEST === `0`)
+      return definition.default;
 
-  async activatePackageManager(locator: Locator) {
-    let emptyFile = false;
-    const lastKnownGoodFile = await getLastKnownGoodFile(`r+`).catch(err => {
-      if ((err as NodeError)?.code === `ENOENT`) {
-        emptyFile = true;
-        return getLastKnownGoodFile(`w`);
-      }
+    const reference = await corepackUtils.fetchLatestStableVersion(definition.fetchLatestFrom);
 
-      throw err;
-    });
     try {
-      await activatePackageManagerFromFileHandle(lastKnownGoodFile, emptyFile || await getJSONFileContent(lastKnownGoodFile), locator);
-    } finally {
-      await lastKnownGoodFile.close();
+      await activatePackageManager(lastKnownGood, {
+        name: packageManager,
+        reference,
+      });
+    } catch {
+      // If for some reason, we cannot update the last known good file, we
+      // ignore the error.
     }
+
+    return reference;
+  }
+
+  async activatePackageManager(locator: Locator) {
+    const lastKnownGood = await getLastKnownGood();
+    await activatePackageManager(lastKnownGood, locator);
```

Comment:
This pattern of reading a file, closing it, then writing to overwrite its content can result in race conditions if there are two processes trying to do that operation in parallel. That's why I refactored the code to use `FileHandle`s in https://github.com/nodejs/corepack/pull/364. That being said, maybe supporting ro FS is more important than supporting running two `corepack` processes in parallel.

There's a failing test ("should support running globally") to address.